pkg/cmd/roachtest/tests/mixed_version_change_replicas.go

```diff
@@ -253,18 +279,82 @@ func runChangeReplicasMixedVersion(ctx context.Context, t test.Test, c cluster.C
 	}
 
 	// Set up and run test.
-	mvt := mixedversion.NewTest(ctx, t, t.L(), c, c.All(), mixedversion.ClusterSettingOption(
-		// Speed up the queues.
-		install.EnvOption{"COCKROACH_SCAN_MAX_IDLE_TIME=10ms"}),
+	mvt := mixedversion.NewTest(
+		ctx, t, t.L(), c, c.All(), mixedversion.ClusterSettingOption(
+			// Speed up the queues.
+			install.EnvOption{"COCKROACH_SCAN_MAX_IDLE_TIME=10ms"},
+		),
 		// Avoid repeatedly running into #114549 on earlier minor versions.
 		// TODO(kvoli): Remove in 24.2.
 		mixedversion.AlwaysUseLatestPredecessors,
-		// Multi-tenant deployments are currently unsupported. See #127378.
-		mixedversion.EnabledDeploymentModes(mixedversion.SystemOnlyDeployment),
 	)
 
 	mvt.OnStartup("create test table", createTable)
 	mvt.InMixedVersion("move replicas", moveReplicas)
 	mvt.AfterUpgradeFinalized("move replicas", moveReplicas)
 	mvt.Run()
 }
+
+// enableTenantSplitScatter updates cluster settings that allow
+// tenants to use SPLIT AT and SCATTER. This is only performed if the
+// mixedversion test is running on a multitenant deployment, and only
+// if required by the active version.
+func enableTenantSplitScatter(l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+	settings := []string{
+		"sql.split_at.allow_for_secondary_tenant.enabled",
+		"sql.scatter.allow_for_secondary_tenant.enabled",
+	}
+
+	for _, s := range settings {
+		// Only enable the relevant settings if they are not already
+		// enabled by default.
+		if !h.Context().FromVersion.AtLeast(mixedversion.TenantsAndSystemAlignedSettingsVersion) {
+			if err := setTenantSetting(l, r, h, s, true); err != nil {
+				return errors.Wrapf(err, "failed to set cluster setting %s", s)
+			}
+		}
+	}
+
+	return nil
+}
+
+// setTenantSetting sets the cluster setting of the given name on
+// the tenant created in for the mixedversion test. After setting it
+// via the system tenant, it also waits until the update is visible to
+// the actual tenant, making sure that statements that need the
+// setting to be enabled can run successfully.
+//
+// It is a no-op to call this function on single tenant (system-only)
+// deployments.
+func setTenantSetting(
+	l *logger.Logger, r *rand.Rand, h *mixedversion.Helper, name string, value bool,
+) error {
+	if !h.IsMultitenant() {
+		return nil
+	}
+
+	if err := h.System.Exec(
+		r,
+		fmt.Sprintf(`ALTER TENANT $1 SET CLUSTER SETTING %s = $2`, name),
+		h.Tenant.Descriptor.Name, value,
+	); err != nil {
+		return errors.Wrapf(err, "failed to set %s", name)
+	}
+
+	return testutils.SucceedsSoonError(func() error {
+		var currentValue bool
+		if err := h.QueryRow(r, fmt.Sprintf("SHOW CLUSTER SETTING %s", name)).Scan(&currentValue); err != nil {
+			return errors.Wrapf(err, "failed to retrieve setting %s", name)
+		}
+
+		if currentValue != value {
+			err := fmt.Errorf(
+				"waiting for setting %s: current (%t) != expected (%t)", name, currentValue, value,
+			)
+			l.Printf("%v", err)
```

Comment:
Is this log print intentional to display interim state?

Reply:
Yeah, it's intentional. If this is actually retried (especially more than once), I think it will provide important context in debugging failures.